knora/dsplib/utils/xml_upload.py

```diff
@@ -910,94 +918,340 @@ def xml_upload(input_file: str, server: str, user: str, password: str, imgdir: s
+                        }
+                    },
+                    "@context": existing_resource['@context']
+                }
+                jsondata = json.dumps(jsonobj, indent=4, separators=(',', ': '), cls=KnoraStandoffXmlEncoder)
+
+                # execute API call
+                answer = None
+                for _ in range(5):
+                    try:
+                        answer = con.put(path='/v2/values', jsondata=jsondata)
+                        stashed_xml_texts[resource][link_prop].pop(pure_text)
+                        break
+                    except ConnectionError:
+                        print(f'{datetime.now().isoformat()}: Try reconnecting to DSP server...')
+                        time.sleep(1)
+                        continue
+                    except RequestException:
+                        print(f'{datetime.now().isoformat()}: Try reconnecting to DSP server...')
+                        time.sleep(1)
+                        continue
+                    except BaseError:
+                        print(f'ERROR while updating the xml text of "{link_prop.name}" of resource "{resource.id}"')
+                        break
+                if verbose:
+                    if answer:
+                        print(f'  Successfully updated xml text of "{link_prop.name}"\n')
 
-    # update the resources with the stashed resptrs
-    if len(stashed_resptr_props) > 0:
-        print('Update the stashed resptrs...')
-    for resource, prop_2_resptrs in stashed_resptr_props.items():
+    # make a purged version of stashed_xml_texts, without empty entries
+    nonapplied_xml_texts: dict[XMLResource, dict[XMLProperty, dict[str, KnoraStandoffXml]]] = {}
+    for res, propdict in stashed_xml_texts.items():
+        for prop, xmldict in propdict.items():
+            if len(xmldict) > 0:
+                if res not in nonapplied_xml_texts:
+                    nonapplied_xml_texts[res] = {}
+                nonapplied_xml_texts[res][prop] = xmldict
+
+    return nonapplied_xml_texts
+
+
+def update_stashed_resptr_props(
+    verbose: bool,
+    id2iri_mapping: dict[str, str],
+    con: Connection,
+    stashed_resptr_props: dict[XMLResource, dict[XMLProperty, list[str]]]
+) -> dict[XMLResource, dict[XMLProperty, list[str]]]:
+    """
+    After all resources are uploaded, the stashed resptr props must be applied to their resources in knora.
+
+    Args:
+        verbose: bool
+        id2iri_mapping: mapping of ids from the XML file to IRIs in knora
+        con: connection to knora
+        stashed_resptr_props: all resptr props that have been stashed
+
+    Returns:
+        nonapplied_resptr_props: the resptr props that could not be updated
+    """
+
+    print('Update the stashed resptrs...')
+    for resource, prop_2_resptrs in stashed_resptr_props.copy().items():
         print(f'Update resptrs of resource "{resource.id}"...')
```

Comment:
Shouldn't this be printed only if verbose=True? I'm not sure, but I feel that this could be very verbose in case of lots of resptrs..?